File: src/main/java/no/nav/tjenestepensjon/simulering/service/StillingsprosentServiceImpl.java
package no.nav.tjenestepensjon.simulering.service;

import static no.nav.tjenestepensjon.simulering.AppMetrics.Metrics.APP_NAME;
import static no.nav.tjenestepensjon.simulering.AppMetrics.Metrics.APP_TOTAL_STILLINGSPROSENT_CALLS;
import static no.nav.tjenestepensjon.simulering.AppMetrics.Metrics.APP_TOTAL_STILLINGSPROSENT_TIME;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import no.nav.tjenestepensjon.simulering.AppMetrics;
import no.nav.tjenestepensjon.simulering.AsyncExecutor;
import no.nav.tjenestepensjon.simulering.AsyncExecutor.AsyncResponse;
import no.nav.tjenestepensjon.simulering.StillingsprosentCallable;
import no.nav.tjenestepensjon.simulering.TjenestepensjonsimuleringEndpointRouter;
import no.nav.tjenestepensjon.simulering.domain.Stillingsprosent;
import no.nav.tjenestepensjon.simulering.domain.TPOrdning;
import no.nav.tjenestepensjon.simulering.domain.TpLeverandor;
import no.nav.tjenestepensjon.simulering.exceptions.DuplicateStillingsprosentEndDateException;
import no.nav.tjenestepensjon.simulering.exceptions.MissingStillingsprosentException;

@Component
public class StillingsprosentServiceImpl implements StillingsprosentService {

    private static final Logger LOG = LoggerFactory.getLogger(Stillingsprosent.class);
    private final AsyncExecutor<List<Stillingsprosent>, StillingsprosentCallable> asyncExecutor;
    private final TjenestepensjonsimuleringEndpointRouter simuleringEndPointRouter;
    private final AppMetrics metrics;

    public StillingsprosentServiceImpl(AsyncExecutor<List<Stillingsprosent>, StillingsprosentCallable> asyncExecutor,
            TjenestepensjonsimuleringEndpointRouter simuleringEndPointRouter,
            AppMetrics metrics) {
        this.asyncExecutor = asyncExecutor;
        this.simuleringEndPointRouter = simuleringEndPointRouter;
        this.metrics = metrics;
    }

    @Override
    public StillingsprosentResponse getStillingsprosentListe(String fnr, Map<TPOrdning, TpLeverandor> tpOrdningAndLeverandorMap) {
        Map<TPOrdning, StillingsprosentCallable> callableMap = toCallableMap(fnr, tpOrdningAndLeverandorMap);
        metrics.incrementCounter(APP_NAME, APP_TOTAL_STILLINGSPROSENT_CALLS);
        long startTime = metrics.startTime();
        AsyncResponse<TPOrdning, List<Stillingsprosent>> asyncResponse = asyncExecutor.executeAsync(callableMap);
        long elapsed = metrics.elapsedSince(startTime);
        LOG.info("Retrieved all stillingsprosenter in: {} ms", elapsed);
        metrics.incrementCounter(APP_NAME, APP_TOTAL_STILLINGSPROSENT_TIME, elapsed);
        return new StillingsprosentResponse(asyncResponse.getResultMap(), asyncResponse.getExceptions());
    }

    @Override
    public TPOrdning getLatestFromStillingsprosent(Map<TPOrdning, List<Stillingsprosent>> map) throws DuplicateStillingsprosentEndDateException, MissingStillingsprosentException {
        TPOrdning latestOrdning = null;
        Stillingsprosent latestPct = null;
        for (Map.Entry<TPOrdning, List<Stillingsprosent>> entry : map.entrySet()) {
            for (Stillingsprosent stillingsprosent : entry.getValue()) {
                LOG.info("TPORDNING {} STILLINGSPROSENT {}", entry.getKey(), stillingsprosent);
                if (latestOrdning == null) {
                    latestOrdning = entry.getKey();
                    latestPct = stillingsprosent;
                } else {
                    Stillingsprosent pct = getLatest(latestPct, stillingsprosent);
                    if (!pct.equals(latestPct)) {
                        latestOrdning = entry.getKey();
                        latestPct = stillingsprosent;
                    }
                }
            }
        }
        if (latestPct == null) {
            throw new MissingStillingsprosentException("Could not find any stillingsprosent");
        }
        return latestOrdning;
    }

    private Stillingsprosent getLatest(Stillingsprosent latest, Stillingsprosent other) throws DuplicateStillingsprosentEndDateException {
        if (latest.getDatoTom() == null && other.getDatoTom() != null) {
            return latest;
        } else if (latest.getDatoTom() != null && other.getDatoTom() == null) {
            return other;
        } else if (latest.getDatoTom() == null && other.getDatoTom() == null || latest.getDatoTom().isEqual(other.getDatoTom())) {
            throw new DuplicateStillingsprosentEndDateException("Could not decide latest stillingprosent due to multiple stillingsprosent having the same end date");
        } else if (latest.getDatoTom() == null || latest.getDatoTom().isAfter(other.getDatoTom())) {
            return latest;
        } else if (other.getDatoTom() == null || other.getDatoTom().isAfter(latest.getDatoTom())) {
            return other;
        } else {
            return latest;
        }
    }

    private Map<TPOrdning, StillingsprosentCallable> toCallableMap(String fnr, Map<TPOrdning, TpLeverandor> tpOrdningAndLeverandorMap) {
        Map<TPOrdning, StillingsprosentCallable> callableMap = new HashMap<>();
        tpOrdningAndLeverandorMap.keySet()
                .forEach(tpOrdning -> callableMap.put(tpOrdning, new StillingsprosentCallable(fnr, tpOrdning, tpOrdningAndLeverandorMap.get(tpOrdning), simuleringEndPointRouter)));
        return callableMap;
    }
}

File: src/main/kotlin/no/nav/tjenestepensjon/simulering/service/StillingsprosentServiceImpl.kt
package no.nav.tjenestepensjon.simulering.service

import no.nav.tjenestepensjon.simulering.AppMetrics
import no.nav.tjenestepensjon.simulering.AsyncExecutor
import no.nav.tjenestepensjon.simulering.StillingsprosentCallable
import no.nav.tjenestepensjon.simulering.TjenestepensjonsimuleringEndpointRouter
import no.nav.tjenestepensjon.simulering.exceptions.DuplicateStillingsprosentEndDateException
import no.nav.tjenestepensjon.simulering.exceptions.MissingStillingsprosentException
import no.nav.tjenestepensjon.simulering.model.v1.domain.FNR
import no.nav.tjenestepensjon.simulering.model.v1.domain.Stillingsprosent
import no.nav.tjenestepensjon.simulering.model.v1.domain.TPOrdning
import no.nav.tjenestepensjon.simulering.TPOrdningStillingsprosentMap
import no.nav.tjenestepensjon.simulering.TPOrdningTpLeverandorMap
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class StillingsprosentServiceImpl(
        private val asyncExecutor: AsyncExecutor<List<Stillingsprosent>, StillingsprosentCallable>,
        private val simuleringEndPointRouter: TjenestepensjonsimuleringEndpointRouter,
        private val metrics: AppMetrics) : StillingsprosentService
{

    override fun getStillingsprosentListe(fnr: FNR, tpOrdningAndLeverandorMap: TPOrdningTpLeverandorMap): StillingsprosentResponse {
        val callableMap = toCallableMap(fnr, tpOrdningAndLeverandorMap)
        metrics.incrementCounter(AppMetrics.Metrics.APP_NAME, AppMetrics.Metrics.APP_TOTAL_STILLINGSPROSENT_CALLS)
        val startTime = metrics.startTime()
        val asyncResponse = asyncExecutor.executeAsync(callableMap)
        val elapsed = metrics.elapsedSince(startTime)
        LOG.info("Retrieved all stillingsprosenter in: {} ms", elapsed)
        metrics.incrementCounter(AppMetrics.Metrics.APP_NAME, AppMetrics.Metrics.APP_TOTAL_STILLINGSPROSENT_TIME, elapsed.toDouble())
        return StillingsprosentResponse(asyncResponse.resultMap, asyncResponse.exceptions)
    }

    @Throws(DuplicateStillingsprosentEndDateException::class,
            MissingStillingsprosentException::class)
    override fun getLatestFromStillingsprosent(map: TPOrdningStillingsprosentMap) =
            map.flatMap { (key, list) ->
                list.map { value ->
                    LOG.info("TPORDNING {} STILLINGSPROSENT {}", key, value)
                    key to value
                }
            }.ifEmpty { throw MissingStillingsprosentException("Could not find any stillingsprosent") }
            .reduce(::getLatest).first

    @Throws(DuplicateStillingsprosentEndDateException::class)
    private fun getLatest(latest: Pair<TPOrdning, Stillingsprosent>, other: Pair<TPOrdning, Stillingsprosent>) = when {
            latest.second.datoTom == other.second.datoTom -> throw DuplicateStillingsprosentEndDateException("Could not decide latest stillingprosent due to multiple stillingsprosent having the same end date")
            other.second.datoTom == null -> other
            latest.second.datoTom == null || latest.second.datoTom!! > other.second.datoTom -> latest
            latest.second.datoTom!! < other.second.datoTom -> other
            else -> latest
        }

    private fun toCallableMap(fnr: FNR, tpOrdningAndLeverandorMap: TPOrdningTpLeverandorMap) =
            tpOrdningAndLeverandorMap.map { (tpOrdning, tpLeverandor) ->
                tpOrdning to StillingsprosentCallable(fnr, tpOrdning, tpLeverandor, simuleringEndPointRouter)
            }.toMap()

    companion object {
        @JvmStatic
        @Suppress("JAVA_CLASS_ON_COMPANION")
        private val LOG = LoggerFactory.getLogger(javaClass.enclosingClass)
    }

}